netutils/config/parser.py

```diff
@@ -1419,3 +1421,69 @@ def _build_banner(self, config_line: str) -> t.Optional[str]:
                 except StopIteration:
                     return None
         raise ValueError("Unable to parse banner (system note) end.")
+
+
+class PaloAltoNetworksConfigParser(BaseSpaceConfigParser):
+    """Palo Alto Networks config parser."""
+
+    comment_chars: t.List[str] = []
+    banner_start: t.List[str] = ['set system login-banner "']
+
+    @property
+    def banner_end(self) -> str:
+        """Demarcate End of Banner char(s)."""
+        raise NotImplementedError("Palo Alto Networks platform doesn't have a banner.")
+
+    def build_config_relationship(self) -> t.List[ConfigLine]:
+        r"""Parse text of config lines and find their parents.
+
+        Examples:
+            >>> config = (
+            ...     "set devices localhost.localdomain deviceconfig system hostname firewall1\n"
+            ...     "set devices localhost.localdomain deviceconfig system panorama local-panorama panorama-server 10.0.0.1\n"
+            ...     "set devices localhost.localdomain deviceconfig system panorama local-panorama panorama-server-2 10.0.0.2\n"
+            ...     "set devices localhost.localdomain deviceconfig setting config rematch yes\n"
+            ... )
+            >>> config_tree = PaloAltoNetworksConfigParser(config)
+            >>> config_tree.build_config_relationship() == \
+            ... [
+            ...     ConfigLine(config_line="set devices localhost.localdomain deviceconfig system hostname firewall1", parents=()),
+            ...     ConfigLine(config_line="set devices localhost.localdomain deviceconfig system panorama local-panorama panorama-server 10.0.0.1", parents=()),
+            ...     ConfigLine(config_line="set devices localhost.localdomain deviceconfig system panorama local-panorama panorama-server-2 10.0.0.2", parents=()),
+            ...     ConfigLine(config_line="set devices localhost.localdomain deviceconfig setting config rematch yes", parents=()),
+            ... ]
+            True
+        """
+        # if config is in palo/json format, convert to set
+        if self.config_lines_only.startswith("config {") and self.config_lines_only is not None:
+            converted_config = paloalto_panos_brace_to_set(self.generator_config)
```

Comment:
If we are going to normalize to set, why is this based on a brace parser?

Reply:
I'm not following? This would allow panos backups to be in brace or set format, but the class is inheriting from basespaceparser. Set is preferred since, as discussed, brace would require custom compliance methods.

Reply:
as we spoke on slack, my mistake, I got confused with Space vs Brace.